### PR TITLE
ENH: Integrate new infrastructure in NiWorkflows to handle spatial references

### PR DESCRIPTION
**Pull request by mgxd**
upstream PR - poldracklab/fmriprep#1915

### Changes
- change `output_spaces` `OrderedDict` in function signatures to `spaces`, a `Spaces` object.

### Maintenance 
- Remove `--template`, `--fs-output-spaces`, and `--no-freesurfer` command line arguments

### CODE REVIEW

**Comment by mgxd** [comment]
@oesteban this roughly models the behavior discussed in poldracklab/fmriprep#1604, though since we're essentially treating this as a new language (spac-ese?) it would be good to cement the conventions, especially when multiple entities are given.

An example of a potentially ambiguous input:
`--output-spaces MNI152:label1-a:label2-b` 

Should this produce 1 MNI152 output with label1 / label2 specifications, or 2 MNI152 outputs each pertaining to a label spec?

To make this clearer, I propose introducing the `,` character to denote multiple specifications for a single space.

i.e.
* `--output-spaces MNI152NLin2009cAsym:res-1:res-2` should produce 2 resolutions of the MNI152NLin2009cAsym template

* `--output-spaces MNIPediatricAsym:cohort-1,res-2` should produce 1 resolution of cohort 1
* `--output-spaces MNIPediatricAsym:cohort-1:res-2` should produce the default (unspecified) resolution of cohort-1, and a lower resolution default MNIPediatricAsym (if there is one available)

WDYT?

File: smriprep/cli/run.py [hunk 1]
```diff
@@ -82,10 +82,6 @@ def get_parser():
     g_conf.add_argument(
         '--longitudinal', action='store_true',
         help='treat dataset as longitudinal - may increase runtime')
-    g_conf.add_argument(
```
Comment (oesteban) [inline smriprep/cli/run.py]: thanks for removing this.

File: smriprep/cli/run.py [hunk 2]
```diff
@@ -294,38 +278,23 @@ def build_workflow(opts, retval):
     from nipype import logging, config as ncfg
     from ..__about__ import __version__
     from ..workflows.base import init_smriprep_wf
+    from ..utils import Spaces
     from niworkflows.utils.bids import collect_participants
 
-    # Set the default template to 'MNI152NLin2009c'
-    output_spaces = opts.output_spaces or OrderedDict([('MNI152NLin2009cAsym', {})])
+    spaces = Spaces(output=opts.output_spaces)
```
Comment (oesteban) [inline smriprep/cli/run.py]: Will write a PR to your branch integrating the solution I have just proposed in NiWorkflows

File: smriprep/cli/run.py [hunk 3]
```diff
@@ -294,38 +278,23 @@ def build_workflow(opts, retval):
...
+    target_spaces = spaces.unique()
+    if not target_spaces:
+        # no output spaces have been set - should this blow up?
```
Comment (oesteban) [inline smriprep/cli/run.py]: sMRIPrep should not add any space, and should not blow up. 

fMRIPrep will always add MNI152NLin2009cAsym (hence, it shouldn't blow up either)

File: smriprep/cli/run.py [hunk 4]
```diff
@@ -294,38 +278,23 @@ def build_workflow(opts, retval):
...
-    FS_SPACES = set(['fsnative', 'fsaverage', 'fsaverage6', 'fsaverage5'])
-    if opts.run_reconall and not list(FS_SPACES.intersection(output_spaces.keys())):
+    FS_SPACES = {'fsnative', 'fsaverage', 'fsaverage6', 'fsaverage5'}
+    if opts.run_reconall and not FS_SPACES.intersection(target_spaces):
         print("""\
 Although ``--fs-no-reconall`` was not set, no FreeSurfer output space (valid values are: \
 %s) was selected. Adding default "fsaverage5" to the \
 list of output spaces.""" % ', '.join(FS_SPACES), file=sys.stderr)
-        output_spaces['fsaverage5'] = {}
+        spaces.add_space('fsaverage5', output=False)
```
Comment (oesteban) [inline smriprep/cli/run.py]: Okay, you were right we have a problem here with my proposal. I'll mull over it tonight - I kind of like your approach.

File: smriprep/workflows/norm.py [hunk 5]
```diff
@@ -200,8 +200,9 @@ def init_anat_norm_wf(
     ])
 
     # Provide synchronized output
-    outputnode = pe.JoinNode(niu.IdentityInterface(fields=out_fields),
-                             name='outputnode', joinsource='split_desc')
+    outputnode = pe.JoinNode(niu.IdentityInterface(fields=out_fields + ['template_and_spec']),
```
Comment (oesteban) [inline smriprep/workflows/norm.py]: This should not be necessary - will propose a solution.

**Comment by oesteban** [comment]
Please check out mgxd/smriprep#2

File: smriprep/workflows/anatomical.py [hunk 6]
```diff
@@ -29,9 +29,20 @@
...
+        longitudinal,
+        num_t1w,
+        omp_nthreads,
+        output_dir,
+        reportlets_dir,
+        skull_strip_template,
+        spaces,
+        debug=False,
+        name='anat_preproc_wf',
+        skull_strip_fixed_seed=False
```
Comment (effigies) [inline smriprep/workflows/anatomical.py]: Just a note that we've apparently abandoned our usual rule of no default arguments (except `name`) in workflow building functions.

We might want to think about either re-enforcing it or making an explicit statement in a style guide of when default arguments should be used.

Comment (oesteban) [reply smriprep/workflows/anatomical.py]: Sounds reasonable

**Comment by oesteban** [comment]
@mgxd why blocked?

**Comment by oesteban** [comment]
DS005 needs a cache wipeout bc it's picking up an old normalization reportlet (it should come out without normalization)

**Comment by mgxd** [comment]
need to do some testing on `fmriprep` side

**Comment by oesteban** [comment]
You are referring to poldracklab/fmriprep#1915, right?

File: smriprep/workflows/base.py [hunk 7]
```diff
@@ -209,50 +204,44 @@ def init_single_subject_wf(
...
         Directory in which to save reportlets
-    skull_strip_fixed_seed : bool
+    skull_strip_fixed_seed : :obj:`bool`
         Do not use a random seed for skull-stripping - will ensure
         run-to-run replicability when used with --omp-nthreads 1
-    skull_strip_template : tuple
-        Name of ANTs skull-stripping template (e.g., 'OASIS30ANTs') and
-        dictionary of template specifications.
-    subject_id : str
+    skull_strip_template : :py:class:`~niworkflows.utils.spaces.Space`
```
Comment (oesteban) [inline smriprep/workflows/base.py]: btw, intersphinx is working with niworkflows :)

**Comment by mgxd** [comment]
i think this is fine to merge in - let's try to hold off release until poldracklab/fmriprep#1915 is resolved

**Pull request merged**